orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc

```diff
@@ -0,0 +1,550 @@
+    auto& node = *graph.GetNode(node_index);
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    if (IsATenEmbedding(&node) &&
+        graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders()) &&
+        node.InputDefs().size() >= 3 &&
+        node.InputDefs()[2]->Exists() &&
+        graph_utils::IsConstantInitializer(graph, node.InputDefs()[2]->Name()) &&
+        node.InputDefs()[1]->Exists() &&
+        graph_utils::IsGraphInput(graph, node.InputDefs()[1]) &&
+        node.InputDefs()[1]->Shape()->dim(0).has_dim_param() &&
+        node.InputDefs()[1]->Shape()->dim(1).has_dim_param()) {
+      const TensorProto* tensor_proto = graph_utils::GetConstantInitializer(graph, node.InputDefs()[2]->Name());
+      if (tensor_proto != nullptr &&
+          tensor_proto->dims_size() == 0 &&
+          ((tensor_proto->data_type() == ONNX_NAMESPACE::TensorProto_DataType_INT32) ||
+           (tensor_proto->data_type() == ONNX_NAMESPACE::TensorProto_DataType_INT64))) {
+        int64_t padding_idx = *reinterpret_cast<const int64_t*>(tensor_proto->raw_data().data());
+        if (padding_idx < 0) {
+          continue;
+        }
+        if (std::find(sparse_embedding_input_names_.begin(), sparse_embedding_input_names_.end(),
+                      node.InputDefs()[1]->Name()) == sparse_embedding_input_names_.end()) {
+          LOG_DEBUG_INFO(logger, "Skip node " + node.Name() + "(" + node.OpType() +
+                                     ") due to embedding input is not in the sparse embedding input list.");
+          continue;
+        }
+        embedding_node = &node;
+        input_ids_arg = embedding_node->MutableInputDefs()[1];
+        PushAllOutputNode(graph, to_visit, embedding_node, visited);
+        for (auto output_edges : embedding_node->MutableOutputDefs()) {
+          candidate_edges.insert(output_edges);
+        }
+        break;
+      }
+    }
+  }
+
+  if (!embedding_node) {
+    LOG_DEBUG_INFO(logger, "Exit PaddingElimination optimization for not finding any valid embedding node.");
+    return Status::OK();
+  }
+
+  while (!to_visit.empty()) {
+    Node* cur = to_visit.front();
+    to_visit.pop();
+    visited.insert(cur);
+    if (graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "Add", {7, 13, 14})) {
+      ORT_ENFORCE(candidate_edges.find(cur->MutableInputDefs()[0]) != candidate_edges.end() ||
+                  candidate_edges.find(cur->MutableInputDefs()[1]) != candidate_edges.end());
+      if (candidate_edges.find(cur->MutableInputDefs()[0]) != candidate_edges.end()) {
+        // Shape of [batch_size, seqlen, ...] is propagated from the first argument of cur node.
+        // Now only support the target two dims are absolutely same or the other shape dim size is smaller by 2.
+        // TODO: support other case such as the other shape is [batch_size, 1, ...]
+        if (cur->InputDefs()[1]->Shape()->dim_size() < cur->InputDefs()[0]->Shape()->dim_size() - 1 ||
```

Comment:
I think this condition pose a strong assumption: we are doing the first two dimension merge. If so, we should at the beginning make it clear PaddingElimination target to merge the first two dimension and remove padding on this merged dimension.

Reply:
Yes, I declared it in the 'padding_elimination.h' .